gax-httpjson/src/main/java/com/google/api/gax/httpjson/ApiMessage.java

```diff
@@ -37,9 +37,12 @@
 
 @BetaApi
 public interface ApiMessage {
+  /* For a each fieldName in a set of fieldNames, fetch that field's List<String> value. */
   Map<String, List<String>> populateFieldsInMap(Set<String> fieldNames);
 
   @Nullable
   /* If this is a Request object, return the inner ApiMessage that represents the body of the request; else return null. */
-  ApiMessage getRequestBody();
+  ApiMessage getApiMessageRequestBody();
+
+  Map<String, String> getApiMessagePathParams();
```

Comment:
I'm having trouble seeing how everything fits together now, with an `ApiMessage` knowing how its data maps to path params. The original design was for a message to just be a model, and the method descriptor would know how to map that data to the correct path, query, and body components.

Reply:
Yeah, this was slowly drifting away from ApiMessage being a simple struct. 

Pushing the handling of ResourceName out into the ApiMessageDescriptor and formatter object.